\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{setspace}
\doublespacing
\usepackage{color}
\usepackage{paralist}
\usepackage[round]{natbib}
\bibliographystyle{apalike}
\usepackage{xspace}
\newcommand{\ps}{phylesystem\xspace}
\newcommand{\otol}{Open Tree of Life\xspace}

%opening
\title{Phylesystem: a git-based data store for community curated phylogenetic estimates}
\author{Emily Jane McTavish, Mark T. Holder ... ??}

\usepackage{hyperref}
\hypersetup{colorlinks=true, linkcolor=red, citecolor=red, hyperindex=true, backref}
\begin{document}

\maketitle
Target: Bioinformatics\\
Scope: Databases and ontologies\\
Type: Original Papers (up to 7 pages; this is approx. 5000 words).
Original papers that describe new research developments in computational molecular biology, for example: 
models, algorithms, software involving new methods, biological databases and network information services, and their impact on molecular biology or computer science. 
Actual biological data, as opposed to purely simulated data, must be used.


\begin{abstract}
    {\em Motivation}:
Phylogenetic relationships must be taken into account when conducting any analysis that compares more than two species.
TreeBase provides archival storage of the data that were used to infer many phylogenies and the phylogenetic statements.
However, available trees (both those in TreeBase many which have not been submitted to TreeBase) frequently benefit from some editing
 to improve the accuracy and reusability of the phylogenetic statements.
For example, many trees available in supplemental data files are incorrectly rooted.
Establishing the linkage between tip labels used in a study and taxa in a unified taxonomy dramatically improves the
    ability of other researchers to reuse phylogenetic estimates.
The Open Tree of Life project has begun curating phylogenetic estimates in an attempt to produce into a unified estimate of the tree of life.
That project also aims to provide infrastructure to support the open curation of phylogenetic data by the community of biologists.
Because process of curating a published phylogenetic estimate is not error-free, storing a full record of the provenance of the information
associated with at tree is crucial for openness and quality assurance.\\
{\em Results}:
Here we describe a software system for providing an interface for the standard database operations of creating, reading, updating, and deleting
phylogenetic statements around a git repository.
The record of the history of edits to a tree is preserved by git's version control features.
Hosting this datastore on GitHub provides a open access to the data store using tools familiar to many developers.
We have developed a ``phylesystem-api'' to implement wrappers around git, interaction with GitHub, and to return readable diffs for complex data structures. \\
{\em Availability}:
Source code for the web-service layer is available at \url{https://github.com/OpenTreeOfLife/phylesystem-api}
and \url{https://github.com/OpenTreeOfLife/peyotl}.
The git data store is available from \url{https://github.com/OpenTreeOfLife/phylesystem}
\end{abstract}

\section{DRAFT}
\subsection{Introduction}
Characterizing and systematizing relationships among species has been a goal of biologists since Linnaeus \cite{Linne1758}.
The phylogenetic systematics ``revolution'' of the 1960's focused most of the effort toward this goal around the 
    task of estimating phylogenetic relationships as the most important tool for classifying life on Earth.
The appreciation of the importance of phylogenetically-aware comparative methods \citep[e.g.][]{Felsenstein1985Comp} and 
the rapid growth in availability of molecular data have led to dramatic increases in the number phylogenetic estimates.

Since 1994, TreeBase \citep{SandersonDPE1994} and its successor TreeBase v.~2 have served
    as the primary archival data stores for phylogenetic estimates and the data that these trees are based on.
As of 2014, TreeBase contains 4076 studes and over 12,800 trees \citep{TreeBaseWebCite}.
Many journals encourage or require that each publications that estimate phylogenetic trees create a TreeBase
    deposition for the publication.

Unfortunately, the current state of interoperabilty in ``phyloinformatics'' makes it difficult and error prone
to capture the inputs and outputs of a phylogenetic analysis in a rich, uniform manner.
Some data stores have reacted to this challenge by accepting a wide range of inputs and making few or no
    guarantees to users of the database that the records will be in any particular form.
On the opposite end of the spectrum, some databases require data providers to conform to more rigorous
    standards with respect to input format and content.

Generic archiving services, such as datadryad\citep{Dryad}, emphasize open access to data and easy of submission by
    accepting almost any format of data deposit.
If the author's submitting the data are not conscientious in their explanations of the data, it can be difficult
    for users of the data to extract all of the structure.

TreeBase produces a much more feature-rich interface because each submission is parsed thoroughly and converted 
    to a set of records in a relational database.
The downside of this approach is that it requires that submissions correspond to a more uniform format.
If there were a universally used, rich file format for phylogenetics analyses, then the restriction would not be a
    serious hurdle.
Unfortunately, file formats read and written by tree estimation tools are idiosyncratic and often so 
    terse that they omit useful information about the analysis.
Thus the TreeBase submission process usually requires the authors of a data package to reformat their data, correct
    the rooting of the tree, alter the labels of the tips of the tree, {\em etc}.
The TreeBase project currently provides an approximately 12-minute video explaing the procedure at 
    \url{http://treebase.org/treebase-web/submitTutorial.html}.
Because these steps are often taken after the analyses for a publication are complete, it is more likely 
    errors introduced in the submission process will not be corrected until an observant user tries to reuse the data.

Unfortunately, despite recent developments in data sharing and digital repositories,
    the vast majorities of phylogenies are not digitally available \cite{Drew}.

The state of phyloinformatic archiving described above suggests the need for a datastore that allows the 
    community of biologists to improve the accuracy and re-usability of published phylogenetic statements.
We describe here an attempt at one such system, which we call ``phylesystem'' because it uses a set of
    versioned text files on the filesystem (and the fact that we {\em love} bad puns).
The goal of \ps is not to replace systems like Dryad or TreeBase, but to complement them by 
    providing phylogenetic statements in consistent format while retaining the history of
    edits that were made to the raw data.
We anticipate that most of the trees stored in \ps will be associated with permanent, static archives
    elsewhere (usually either TreeBase or Dryad).

\subsection{Goals}
One of the motivations of the design of \ps was to support the \otol's need for a curated set
    of rooted trees that have been aligned to a common taxonomy.
However, one of the goals of the \otol project was to support infrastructure for phylogenetic
    research in an open way that encourages the community of biologists to participate in the 
    collection and curation of our knowledge of the phylogenetic relationships of life on Earth.

Thus, our goals when designing \ps were to build:
\begin{compactenum}
    \item a data store capable of storing phylogenetic estimates from thousands of published studies; we do not intend to store the data upon which the phylogenetic estimates are based.
    \item a data store expected to store a few exemplar trees from each study; we do not anticipate that the system will be used to store thousands of trees from a bootstrapping or MCMC analysis, for example.
    \item a system capable of storing the full history of changes made to a study and its trees, including an identifier to indicate who made the changes.
    \item a data store rich enough to handle extensive annotation of the trees, the taxa to which they refer to, or metadata describing the analysis that produced the trees.
    \item build a loosely coupled system that would allow the community of biologists in interact with the data in a wide variety of ways.
\end{compactenum}

The relatively small amount of data that we expected to store and the relaitvely small
    number of edits anticipated for each study included in the corpus implied that 
    raw performance of the basic CRUD operations was unlikely to be a bottleneck for most
    uses of the \ps.
Indeed the total size of the stored data (in JSON format with a line ending after 
    every field) is only around 150MB.

There are several requirements for phylogenetic data store.
      - Accessible data store
Accessibility is extremely important.
OpenTree supports open science, and data availability.
For a phylogenetic database to gain adoption it must be readily usable by end users.
Licenses? UGH.
          - Version control
          - OTU mapping
However, making phylogenies re-usable by researchers not involved in the original analyses often requires fairly extensive curation.
The tip labels on trees often have meaning in the context of that analysis that are not easily translated across studies. 
For example, the tip labels may contain abbreviations, or lab sample codes.
Data reuse requires standardizing labels so that they are meaningful across studies.
Taxonomic name recognition services, such as ?? Plant something, or ?? taxomachine, can attempt to automatically map names to
operational taxonomic units, but this process also requires come subjective human intervention. 

          - rerooting
Additionally, many phylogenetic inference programs (e.g. ????) return unrooted phylogenies, 
which researchers then manually root for inclusion in figures or inference. 
If these rootings are not included in the shared tree file, 
these phylogenies need to be manually rooted for inclusion in a synthetic tree or downstream analyses

These curation steps require human input and are therefore somewhat subjective.
A sustainable database requires a mechanism for correcting and updating these trees,
while retaining provenance for individual changes.

Ideally, a widely adopted database could support multiple editors curating contributed phylogenies simultaneously.
Thus a system for merging and combining inputs in necessary.

By supporting contribution and curation in an open and accessible manner, we hope to provide a database with longterm utility and widespread adoption.

          - many curators!, possibly editing studies at the same time!
          - long term used, ideally.

           Implementation of database (structure):
      - FIGURE TO REPRESENT!
      - File Formats
         - Explain Nexson... UGH
         
      - Sharded git repository
          Describe full structure 
          
      - External library to index studies (is OTI described somewhere?)
      - peyotl...
      
      Curation via app:
        - main repo on server -> mirror -> github
            ??? do pulls from github happen? wholes cycle *can* be reversed.
        - dealing with WIP branches.
        *Even conflicts that Git may find mergable"*
        return merge conflicts to user

 Discussion:
   Generalizability:
      Git may be a good option for any non-static data store.
      
  Future directions

  
\bibliography{phylesystem}  



\section{notes}
  --------------------------------------------------------------------
Key Points:
Pros
 - Versioned
 - Familiar to users
 - Comes with a lot of great machinery already enabled

diffing doesn't always behave well.
simple data structures behave oddly...

Meaningful merges
 -- immutable
 -- can't

Is this model applicable for database store

General model based decisions:
- Decisions that studies won't move.

- given the study ID what shard will it be in.

- scales better in general.

- Data is so open.



 The current (as of May) data set
 Community contributed phylogenies

 - 6745 trees from 2914 published studies
 - 1188 trees from 991 studies partly curated 
 - 335 trees from 327 studies completely curated and included in the synthetic tree.

 The problem:
 - Large data set: Thousands of phylogenies, and always growing (hopefully!)
 - Each phylogeny requires some hand curation, often by multiple people
 - Need to be readily accessible, and editable by interested researchers

Curation

 Potential data store options:
 - SQL database
 - Mongo, couchDB
 - git/github


 We chose git!

 - Trees and annotations by study in Nexson format  
(JSON serialization of NeXML)
 - Whole datastore is a git repo!  
 

 Curation
- Work in progress branch is created upon curation  
- If study hasn't been edited by someone else, changes are automatically merged.  
- Otherwise, merged changes are returned to curator to accept or reject
- Updates pushed to GitHub after each commit
 Features
 - Tracking curation attribution  
- Some subjective choices, edits made by many in the community over time

 Curation

 Features
 These trees are the backend for OpenTree showpiece  
 the synthetic tree!
 - but also a useful datastore for other researchers
 - Repo is hosted on GitHub, entire data store can be easily cloned and updated
 - Anyone can easily download all the data!
 
 Features
- Hosting on Github
- Free  
- Familar to  many in the field


 Potential issues:
 - Phylogenies are hard to diff - e.g. rerooting changes everything!
 - Nexson are not a line based format
 - Repo size limits on github


 In the future:

 - Semantic diffs  
 - Pull requests
 
  
Generality:
Is a git-based datastore right for your project?

 - Maybe! Any of the OpenTree software team are happy to chat about pros and cons.

NSF AVATOL \#1208809  


\end{document}
