\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{setspace}
\doublespacing
\usepackage{color} 
\usepackage[round]{natbib}
\bibliographystyle{apalike}

%opening
\title{Phylesystem: a git-based data store for community curated phylogenetic estimates}
\author{Emily Jane McTavish, Mark T. Holder ... ??}

\usepackage{hyperref}
\hypersetup{colorlinks=true, linkcolor=red, citecolor=red, hyperindex=true, backref}
\begin{document}

\maketitle
Target: Bioinformatics\\
Scope: Databases and ontologies\\
Type: Original Papers (up to 7 pages; this is approx. 5000 words).
Original papers that describe new research developments in computational molecular biology, for example: 
models, algorithms, software involving new methods, biological databases and network information services, and their impact on molecular biology or computer science. 
Actual biological data, as opposed to purely simulated data, must be used.


\begin{abstract}
    {\em Motivation}:
Phylogenetic relationships must be taken into account when conducting any analysis that compares more than two species.
TreeBase provides archival storage of the data that were used to infer many phylogenies and the phylogenetic statements.
However, available trees (both those in TreeBase many which have not been submitted to TreeBase) frequently benefit from some editing
 to improve the accuracy and reusability of the phylogenetic statements.
For example, many trees available in supplemental data files are incorrectly rooted.
Establishing the linkage between tip labels used in a study and taxa in a unified taxonomy dramatically improves the
    ability of other researchers to reuse phylogenetic estimates.
The Open Tree of Life project has begun curating phylogenetic estimates in an attempt to produce into a unified estimate of the tree of life.
That project also aims to provide infrastructure to support the open curation of phylogenetic data by the community of biologists.
Because process of curating a published phylogenetic estimate is not error-free, storing a full record of the provenance of the information
associated with at tree is crucial for openness and quality assurance.\\
{\em Results}:
Here we describe a software system for providing an interface for the standard database operations of creating, reading, updating, and deleting
phylogenetic statements around a git repository.
The record of the history of edits to a tree is preserved by git's version control features.
Hosting this datastore on GitHub provides a open access to the data store using tools familiar to many developers.
We have developed a ``phylesystem-api'' to implement wrappers around git, interaction with GitHub, and to return readable diffs for complex data structures. \\
{\em Availability}:
Source code for the web-service layer is available at \url{https://github.com/OpenTreeOfLife/phylesystem-api}
and \url{https://github.com/OpenTreeOfLife/peyotl}.
The git data store is available from \url{https://github.com/OpenTreeOfLife/phylesystem}
\end{abstract}

\section{DRAFT}
\subsection{Introduction}
Characterizing and systematizing relationships among species has been a goal of biologists since Linnaeus \cite{Linne1758}.
The phylogenetic systematics ``revolution'' of the 1960's focused most of the effort toward this goal around the 
    task of estimating phylogenetic relationships as the most important tool for classifying life on Earth.
The appreciation of the importance of phylogenetically-aware comparative methods \citep[e.g.][]{Felsenstein1985Comp} and 
the rapid growth in availability of molecular data have led to dramatic increases in the number phylogenetic estimates.

Recent pushes to understand the evolutionary relationships among species have resulted in a wealth of phylogenetic information. 

In the past 5 years there have been major publications on XXXX angiosperms, ??? bird tree of life, ?? other major trees.
Additionally, much finer resolution studies capturing relationships at a lower phylogenetic level are being published daily.
The information contained in these phylogenies is essential not only for understanding the evolutionary history and systematics of these groups,
but are also an essential framework for studying evolutionary processes, such as trait evolution (REF), dispersal (REF), and speciation (REF)
among many others.

Trees 
     However,
     TOL projects
     Tree base and Phylografter

In order for this data to be reusable for follow-up studies, or inclusion in a synthetic tree, it must be digitally available and accessible \cite{cranston_best_2014}.
Sadly, despite recent developments in data sharing and digital repositories, the vast majorities of phylogenies are not digitally available \cite{Drew}.
Several previous projects have attempted to curate and collate phylogenetic data, most notably TreeBase (REF) and Phylografter (REF).
We have developed a new database for phylogenies which incorporates the trees included ind these previous efforts, 
and builds upon their utility by allowing for community curation and tracking of changes through time, as well as a readily accessible data store.

\subsection{Purpose}
There are several requirements for phylogenetic data store.
      - Accessible data store
Accessibility is extremely important.
OpenTree supports open science, and data availability.
For a phylogenetic database to gain adoption it must be readily usable by end users.
Licenses? UGH.
          - Version control
          - OTU mapping
However, making phylogenies re-usable by researchers not involved in the original analyses often requires fairly extensive curation.
The tip labels on trees often have meaning in the context of that analysis that are not easily translated across studies. 
For example, the tip labels may contain abbreviations, or lab sample codes.
Data reuse requires standardizing labels so that they are meaningful across studies.
Taxonomic name recognition services, such as ?? Plant something, or ?? taxomachine, can attempt to automatically map names to
operational taxonomic units, but this process also requires come subjective human intervention. 

          - rerooting
Additionally, many phylogenetic inference programs (e.g. ????) return unrooted phylogenies, 
which researchers then manually root for inclusion in figures or inference. 
If these rootings are not included in the shared tree file, 
these phylogenies need to be manually rooted for inclusion in a synthetic tree or downstream analyses

These curation steps require human input and are therefore somewhat subjective.
A sustainable database requires a mechanism for correcting and updating these trees,
while retaining provenance for individual changes.

Ideally, a widely adopted database could support multiple editors curating contributed phylogenies simultaneously.
Thus a system for merging and combining inputs in necessary.

By supporting contribution and curation in an open and accessible manner, we hope to provide a database with longterm utility and widespread adoption.

          - many curators!, possibly editing studies at the same time!
          - long term used, ideally.

           Implementation of database (structure):
      - FIGURE TO REPRESENT!
      - File Formats
         - Explain Nexson... UGH
         
      - Sharded git repository
          Describe full structure 
          
      - External library to index studies (is OTI described somewhere?)
      - peyotl...
      
      Curation via app:
        - main repo on server -> mirror -> github
            ??? do pulls from github happen? wholes cycle *can* be reversed.
        - dealing with WIP branches.
        *Even conflicts that Git may find mergable"*
        return merge conflicts to user

 Discussion:
   Generalizability:
      Git may be a good option for any non-static data store.
      
  Future directions

  
\bibliography{phylesystem}  



\section{notes}
  --------------------------------------------------------------------
Key Points:
Pros
 - Versioned
 - Familiar to users
 - Comes with a lot of great machinery already enabled

diffing doesn't always behave well.
simple data structures behave oddly...

Meaningful merges
 -- immutable
 -- can't

Is this model applicable for database store

General model based decisions:
- Decisions that studies won't move.

- given the study ID what shard will it be in.

- scales better in general.

- Data is so open.



 The current (as of May) data set
 Community contributed phylogenies

 - 6745 trees from 2914 published studies
 - 1188 trees from 991 studies partly curated 
 - 335 trees from 327 studies completely curated and included in the synthetic tree.

 The problem:
 - Large data set: Thousands of phylogenies, and always growing (hopefully!)
 - Each phylogeny requires some hand curation, often by multiple people
 - Need to be readily accessible, and editable by interested researchers

Curation

 Potential data store options:
 - SQL database
 - Mongo, couchDB
 - git/github


 We chose git!

 - Trees and annotations by study in Nexson format  
(JSON serialization of NeXML)
 - Whole datastore is a git repo!  
 

 Curation
- Work in progress branch is created upon curation  
- If study hasn't been edited by someone else, changes are automatically merged.  
- Otherwise, merged changes are returned to curator to accept or reject
- Updates pushed to GitHub after each commit
 Features
 - Tracking curation attribution  
- Some subjective choices, edits made by many in the community over time

 Curation

 Features
 These trees are the backend for OpenTree showpiece  
 the synthetic tree!
 - but also a useful datastore for other researchers
 - Repo is hosted on GitHub, entire data store can be easily cloned and updated
 - Anyone can easily download all the data!
 
 Features
- Hosting on Github
- Free  
- Familar to  many in the field


 Potential issues:
 - Phylogenies are hard to diff - e.g. rerooting changes everything!
 - Nexson are not a line based format
 - Repo size limits on github


 In the future:

 - Semantic diffs  
 - Pull requests
 
  
Generality:
Is a git-based datastore right for your project?

 - Maybe! Any of the OpenTree software team are happy to chat about pros and cons.

NSF AVATOL \#1208809  


\end{document}
